src/webapis/WebAPIs.ts

```diff
@@ -0,0 +1,234 @@
+
+                // Now, let's gather more info on the event.
+                // IMPORTANT: The following call will return the event without decyphering it, so we're
+                // not obtaining anything that we couldn't also obtain through a homeserver's Admin API.
+                //
+                // By doing this with the reporterClient, we ensure that this feature of Mjölnir can work
+                // with all Matrix homeservers, rather than just Synapse.
+                event = await reporterClient.doRequest("GET", `/_matrix/client/r0/rooms/${roomId}/event/${eventId}`);
+            }
+            let accusedId: string = event["sender"];
+
+            /*
+            Past this point, the following invariants hold:
+
+            - The reporter is a member of `roomId`.
+            - Event `eventId` did take place in room `roomId`.
+            - The reporter could witness event `eventId` in room `roomId`.
+            - Event `eventId` was reported by user `accusedId`.
+            */
+
+            let { displayname: reporterDisplayName }: { displayname: string } = await this.client.doRequest("GET", `/_matrix/client/r0/profile/${encodeURIComponent(reporterId)}/displayname`);
+            let { displayname: accusedDisplayName }: { displayname: string } = await this.client.doRequest("GET", `/_matrix/client/r0/profile/${encodeURIComponent(accusedId)}/displayname`);
+            let roomAliasOrID = roomId;
+            try {
+                roomAliasOrID = await this.client.getPublishedAlias(roomId);
+            } catch (ex) {
+                // Ignore.
+            }
+            let eventShortcut = `https://matrix.to/#/${encodeURIComponent(roomId)}/${encodeURIComponent(eventId)}`;
+            let roomShortcut = `https://matrix.to/#/${encodeURIComponent(roomAliasOrID)}`;
+            let eventContent;
+            if (event["type"] === "m.room.encrypted") {
+                eventContent = "<encrypted content>";
+            } else {
+                eventContent = JSON.stringify(event["content"], null, 2);
+            }
+
+            let reason = request.body["reason"];
+
+            // We now have all the information we need to produce an abuse report.
+
+            // We need to send the report as html to be able to use spoiler markings.
+            // We build this as dom to be absolutely certain that we're not introducing
+            // any kind of injection within the report.
+            const document = new JSDOM("<body>User <code id='reporter-display-name'></code> (<code id='reporter-id'></code>) reported <a id='event-shortcut'>event <span id='event-id'></span></a> sent by user <b><span id='accused-display-name'></span> (<span id='accused-id'></span>)</b> in <a id='room-shortcut'>room <span id='room-alias-or-id'></span></a>.<div>Event content <span id='event-container'><code id='event-content'></code><span></div><div>Reporter commented: <code id='reason-content'></code></body></div>").window.document;
```

Comment:
Does this have to be all on one line? I agree it is overkill at this stage to create a template for reports but that's probably something to keep in mind later on